src/jira.py

```diff
@@ -0,0 +1,161 @@
+import datetime
+import re
+import requests
+
+from settings import default_orgs
+from src.access import get_access_params
+from src.issue import Issue
+
+
+class JiraBoard:
+
+    def __init__(self, project_key: str):
+        """Create a Project storing all issues belonging to the provided project key"""
+
+        self.url = get_access_params('jira')['options']['server'] + default_orgs['jira']
+        self.token = get_access_params('jira')['api_token']
+        self.headers = {'Authorization': 'Basic ' + self.token}
+
+        self.key = project_key
+        self.issues = dict()
+        self.api_call(0)  # Get information for all issues in the project
+
+    def api_call(self, start):
+        """
+        Make API calls until all results have been retrieved. Jira API responses can be paginated, defaulting to 50
+        results per page, so a new call has to be made until the total is reached.
+
+        :param start: The index in the results to start at. Always call this function with start=0
+        """
+
+        response = requests.get(self.url + 'search?jql=project=%s&startAt=%s' % (self.key, str(start)),
+                                headers=self.headers).json()
+
+        for i in response["issues"]:
+            self.issues[i["key"]] = JiraIssue(response=i)
+
+        if response['total'] >= start + response['maxResults']:  # There could be another page of results
+            self.api_call(start + response['maxResults'])
+
+    def get_all_epics(self):
+        """Search for issues in this project with issuetype=Epic"""
+
+        response = requests.get(self.url + 'search?jql=project%20%3D%20' + self.key + '%20AND%20issuetype%20%3D%20Epic').json()
+
+    def get_epic_issues(self, epic_key):
+
+        response = requests.get("%ssearch?jql=cf[10008]='%s'" % (self.url, epic_key)).json()
+
+
+class JiraIssue(Issue):
+
+    def __init__(self, key: str = None, response: dict = None):
+        """
+        Create an Issue object from an issue key or from a portion of an API response
+
+        :param key: If specified, make an API call searching by this issue key
+        :param response: If specified, don't make a new API call but use this response from an earlier one
+        """
+        super().__init__()  # Initiate the super class, Issue
+
+        self.url = get_access_params('jira')['options']['server'] % (default_orgs['jira'])
+        self.token = self.token = get_access_params('jira')['api_token']
+        self.headers = {'Authorization': 'Basic ' + self.token}
+
+        if key:
+            r = requests.get("%ssearch?jql=id=%s" % (self.url, key), headers=self.headers).json()
```

Comment:
This is using old-style Python string substitution. How about `....get(f'{self.url}search?jql=id={key}'....)`? I saw this is a few other places. Unless `%s` string substitution is needed (and I'm not sure when that's the case) I wouldn't use it.

Reply:
replaced all with new style string formatting